balancer/pickfirst_leaf/pickfirst_leaf.go

```diff
@@ -0,0 +1,640 @@
+
+	// Since we have a new set of addresses, we are again at first pass.
+	b.firstPass = true
+	// If multiple endpoints have the same address, they would use the same
+	// subconn because an AddressMap is used to store subconns.
+	// This would result in attempting to connect to the same subconn multiple times
+	// in the same pass. We don't want this, so we ensure each address is present
+	// in only one endpoint before moving further.
+	newEndpoints = deDupAddresses(newEndpoints)
+
+	// Perform the optional shuffling described in gRFC A62. The shuffling will
+	// change the order of endpoints but not touch the order of the addresses
+	// within each endpoint. - A61
+	if cfg.ShuffleAddressList {
+		newEndpoints = append([]resolver.Endpoint{}, newEndpoints...)
+		internal.ShuffleAddressListForTesting.(func(int, func(int, int)))(len(newEndpoints), func(i, j int) { newEndpoints[i], newEndpoints[j] = newEndpoints[j], newEndpoints[i] })
+	}
+
+	if b.state == connectivity.Ready {
+		// If the previous ready subconn exists in new address list,
+		// keep this connection and don't create new subconns.
+		prevAddr := b.addressIndex.currentAddress()
+		b.addressIndex.updateEndpointList(newEndpoints)
+		if b.addressIndex.seekTo(prevAddr) {
+			return nil
+		}
+	} else {
+		b.addressIndex.updateEndpointList(newEndpoints)
+	}
+
+	// Remove old subConns that were not in new address list.
+	oldAddrs := resolver.NewAddressMap()
+	for _, k := range b.subConns.Keys() {
+		oldAddrs.Set(k, true)
+	}
+
+	// Flatten the new endpoint addresses.
+	newAddrs := resolver.NewAddressMap()
+	for _, endpoint := range newEndpoints {
+		for _, addr := range endpoint.Addresses {
+			newAddrs.Set(addr, true)
+		}
+	}
+
+	// Shut them down and remove them.
+	for _, oldAddr := range oldAddrs.Keys() {
+		if _, ok := newAddrs.Get(oldAddr); ok {
+			continue
+		}
+		val, _ := b.subConns.Get(oldAddr)
+		val.(*scData).subConn.Shutdown()
+		b.subConns.Delete(oldAddr)
+	}
+
+	// If its the first resolver update or the balancer was already READY or
+	// or CONNECTING, enter CONNECTING.
+	// We may be in TRANSIENT_FAILURE due to a previous empty address list,
+	// we should still enter CONNECTING because the sticky TF behaviour mentioned
+	// in A62 applied only when the TRANSIENT_FAILURE is reported dur to connectivity
+	// failures.
+	if b.state == connectivity.Ready || b.state == connectivity.Connecting || oldAddrs.Len() == 0 {
+		// Start connection attempt at first address.
+		b.state = connectivity.Connecting
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Connecting,
+			Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+		})
+		b.requestConnection()
+	} else if b.state == connectivity.Idle {
+		// If this is not the first resolver update and we're in IDLE, remain
+		// in IDLE until the picker is used.
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Idle,
+			Picker:            &idlePicker{exitIdle: b.ExitIdle},
+		})
```

Comment:
Sorry about the back and forth. 

But thinking about it now, it might still make sense to do one of the following:
- Before the `if` condition in `if b.state == connectivity.Ready || b.state == connectivity.Connecting || prevAddrsCount == 0 {...}`, we should probably do `if b.state == connectivity.Idle { return nil }`, or
- At the last line `return nil`, add a comment saying this covers the case for the LB policy being in `Idle` and that we don't have to do anything. The address list has been updated, and the next time we are asked to exit idle, we will start connecting to the new set of addresses.

Either of these approaches will make the handling for `Idle` explicit and won't leave it up to the reader to figure it out by themselves.